### PR TITLE
[RFC] Define and document per-key ordering semantics for runners

### PR DESCRIPTION
**Pull request by pabloem**
There are various use cases that can be implemented on top of Beam and its runners, but that need certain guarantees regarding the ordered transport of data in between execution stages/steps of a Beam Pipeline.

This PR proposes the following:
- Defining the concept of per-key ordered delivery.
- `ValidatesRunner` tests that allow us to verify runner behavior for per-key ordered delivery.
- Adding a row to the Beam Capability Matrix where runner support for this is documented.

With this, Beam users will be able to enable various order-dependent workloads on Beam/Dataflow with confidence that they need.

-----
Results form ValidatesRunner tests:

- Flink streaming VR tests: [Build Scan](https://scans.gradle.com/s/x4whlt2gchcum/tests#:runners:flink:1.13:job-server-validatesportablerunnerstreaming-org.apache.beam.sdk.transforms-perkeyorderingtest)
- Flink Bartch VR tests: [Jenkins output](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch_PR/261/console) - not running any of the new tests, as it does not support key-ordered delivery.
- Dataflow legacy VR tests: [Build Scan](https://scans.gradle.com/s/pcww3rqmcmu24/tests#:runners:google-cloud-dataflow-java-validatesrunnerlegacyworkertest-org.apache.beam.sdk.transforms-perkeyorderingtest)
- Direct VR tests: [Build Scan](https://scans.gradle.com/s/45pjufvncnoq6/tests#:runners:direct-java-validatesrunner-org.apache.beam.sdk.transforms-perkeyorderingtest)
- Samza tests: [Build Scan]( https://scans.gradle.com/s/npxbjepkaliqk/tests#:runners:samza-validatesrunner-org.apache.beam.sdk.transforms-perkeyorderingtest)
--------------------------------------------

r: @apilloud 
r: @kennknowles 

--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by pabloem** [comment]
Run Dataflow ValidatesRunner

**Comment by pabloem** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by pabloem** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by pabloem** [comment]
Run Dataflow ValidatesRunner

**Comment by pabloem** [comment]
Run Java Spark PortableValidatesRunner Streaming

**Comment by pabloem** [comment]
Run Java Spark PortableValidatesRunner Batch

**Comment by pabloem** [comment]
Run Java Samza PortableValidatesRunner

**Comment by pabloem** [comment]
Run Dataflow PortabilityApi ValidatesRunner
--


<br class="Apple-interchange-newline">

**Comment by pabloem** [comment]
Run Dataflow PortabilityApi ValidatesRunner

**Comment by pabloem** [comment]
Run Direct ValidatesRunner Java 11

**Comment by pabloem** [comment]
Run Dataflow PortabilityApi ValidatesRunner Java 11

**Comment by pabloem** [comment]
Run Java Flink PortableValidatesRunner Streaming

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+        // When matched is set to -1, it means that we have met an error, and elements on this
+        // key are not matched anymore - thus we ignore all inputs.
+      } else if (matched < this.perKeyElements.size()
+          && !this.perKeyElements.get(matched).equals(elm.getValue())) {
+        // If we meet this condition, then the order of elements is not what we're expecting.
+        // We mark `matched` as -1, and output a failed ordering.
+        matchedElements.write(-1);
+        receiver.output(KV.of(elm.getKey(), false));
+      } else {
+        assert this.perKeyElements.get(matched).equals(elm.getValue())
```
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java:82]: the case above is denied (`!...`?)

[the remark this answers is not in the document]
Comment (apilloud) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java:82]: Sorry if my comment wasn't clear, here is what I'm thinking: the case above is `mached < this.perKeyElements.size() && !...` the assert triggers on `!...`, so the assert actually triggers on `matched >= this.perKeyElements.size()`.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    @ProcessElement
+    public void verifyOrder(
+        @Element KV<String, T> elm,
+        OutputReceiver<KV<String, Boolean>> receiver,
+        @StateId("matchedElements") ValueState<Integer> matchedElements) {
+      Integer matched = matchedElements.read();
+      matched = matched == null ? 0 : matched;
+      if (matched == -1) {
+        // When matched is set to -1, it means that we have met an error, and elements on this
+        // key are not matched anymore - thus we ignore all inputs.
```
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java:75]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+      } else if (matched < this.perKeyElements.size()
```
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java:77]: done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+  public void testSingleCallOrderingWithShuffle() {
+    // Here we test that the output of a single process call in a DoFn will be output in order
+    List<Integer> perKeyElements =
+        Lists.newArrayList(-8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8);
+    Collections.shuffle(perKeyElements);
+    List<String> allKeys =
+        Lists.newArrayList(1, 2, 3, 4, 5, 6, 7, 8, 9, 10).stream()
+            .map(elm -> String.format("k%s", elm))
+            .collect(Collectors.toList());
+    PCollection<KV<String, Integer>> kvSeeds =
```
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java:109]: hmmm ... I could try to parameterize the test with different types -though IDK if the `Parameterized` JUnit runner integrates well with the ValidatesRunner framework. Do you have a suggestion? Should I try adding a more complex type instead of just ints?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (apilloud) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java:86]: After the latest update you should never be able to hit this assert, you can leave it as is.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (apilloud) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java:109]: I don't think you'll be able to parameterize this, you'll need different test methods for each type.

I believe runners see these as byte[] today, so the behavior here actually depends on coders, so changes in the coder will change the ordering. (Future relational work will likely make some runners aware of the coder contents as well.) The byte coder is the trivial passthrough type, so `Byte` might actually be the best for a basic "does the runner produce deterministic ordering" test.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+      }
+    }
+  }
+
+  @Test
+  @Category({ValidatesRunner.class, UsesPerKeyOrderedDelivery.class})
+  public void testSingleCallOrderingWithShuffle() {
+    // Here we test that the output of a single process call in a DoFn will be output in order
+    List<Integer> perKeyElements =
+        Lists.newArrayList(-8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8);
```
Comment (apilloud) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java:103]: Sorry for the late addition: You might want to test some more extreme values. VarIntCoder is variable sized, so around those size boundaries might be good to test. https://github.com/apache/beam/blob/50fcf55d60e8fa7a8399d63e030c930a2d45402a/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/VarIntCoder.java#L29

Comment (pabloem) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java]: I tried that. All good.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/UsesPerKeyOrderInStage.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.testing;
+
+import org.apache.beam.sdk.annotations.Internal;
+
+/**
+ * Category tag for validation tests which rely on a runner providing per-key ordering in between
+ * transforms in the same stage. Tests tagged with {@link UsesPerKeyOrderInStage} should be run for
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/UsesPerKeyOrderInStage.java:24]: "Stage" is not a concept in the Beam model, unless you also define it elsewhere in this PR. (commenting as I go)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/UsesPerKeyOrderedDelivery.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.testing;
+
+import org.apache.beam.sdk.annotations.Internal;
+
+/**
+ * Category tag for validation tests which rely on a runner providing per-key ordering. Tests tagged
+ * with {@link UsesPerKeyOrderedDelivery} should be run for runners which support key-to-key
+ * ordering of elements across shuffle / stage boundaries.
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/UsesPerKeyOrderedDelivery.java:25]: "shuffle" is also not a model concept. For example runners can choose to move shuffles around so then the ordered delivery might change. Unless you can demonstrate/require that allowable optimizations preserve the ordered delivery you are looking for.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 8]
```diff
@@ -626,7 +627,7 @@ public void process(OutputReceiver<String> r, PipelineOptions options) {
     }
 
     @Test
-    @Category(ValidatesRunner.class)
+    @Category({ValidatesRunner.class, UsesParDoLifecycle.class})
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java:630]: Seems unrelated.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kennknowles) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/PerKeyOrderingTest.java:87]: Java `assert` is, I think, less preferable than a library that generates more elaborate error messages. Also, failures within a user `DoFn` may be swallowed by a runner. Can the tests be described in terms of a dead-letter output and `PAssert`?

**Review by kennknowles — CHANGES_REQUESTED** [review]
I think that per-key-path ordered delivery is subtle enough it requires a pseudo-mathy treatment in a design doc. I might be behind on dev@ threads but want to check that this has happened. Tests are a useful executable "spec" but when we find a bug that wasn't covered by tests, we go back to the math docs to build the new correct test.

**Comment by pabloem** [comment]
Run Website_Stage_GCS PreCommit

File: website/www/site/content/en/documentation/runtime/model.md [hunk 9]
```diff
@@ -77,6 +82,48 @@ element, and having to retry everything if there is a failure. For example, a
...
+
+Consider a single Beam worker processing a series of bundles from the same Beam
+stage, and consider a `PTransform` that outputs data from this Stage into a
+downstream `PCollection`. Finally, consider two events *with the same key*
+emitted in a certain order by this worker (within the same bundle or as part of
+different bundles).
+
+We say that the Beam runner supports **key-ordered delivery** if it guarantees
+that these two events will be observed downstream in the same order,
+independently of the kind of transmission.
```
Comment (je-ik) [inline website/www/site/content/en/documentation/runtime/model.md:125]: Although it seems to be mentioned I think it would be good to restate here explicitly, that this holds if and only if the two PTransforms are directly connected - or at least there is no intermediate grouping transform in between. "A downstream PCollection" in the above paragraph might be interpreted as *any* downstream PCollection, which does not hold.

Comment (kennknowles) [reply website/www/site/content/en/documentation/runtime/model.md:125]: I was thinking about this comment and it occurs to me that you have to have key-limited parallelism in the producer of course. I don't think this is required anywhere in the model (state/timers require key+window limited parallelism) and also not mentioned here.

Comment (pabloem) [reply website/www/site/content/en/documentation/runtime/model.md]: Hi all! Sorry about the delay, but I've finally tried to address these comments. Can you please take a look and LMK what you think>

File: website/www/site/content/en/documentation/runtime/model.md [hunk 10]
```diff
@@ -50,6 +50,11 @@ ways, such as:
     This may allow the runner to avoid serializing elements; instead, the runner
     can just pass the elements in memory.
 
+Passing elements between transforms that are running on the same worker is
```
[the remark this answers is not in the document]
Comment (pabloem) [reply website/www/site/content/en/documentation/runtime/model.md]: done

File: website/www/site/data/capability_matrix.yaml [hunk 11]
```diff
@@ -1704,3 +1704,42 @@ capability-matrix:
               l1:
               l2:
               l3:
+        - name: Key-ordered delivery
+          description: The runner offers guarantees for the order in which elements are passed in between operations. See <a href=''>per-key ordering semantics.</a>
+          values:
+            - class: dataflow
+              l1: "Partially"
+              l2:
+              l3: Dataflow performs different shuffling algorithms for batch and streaming. Dataflow guarantees key-ordered delivery in streaming, though not in batch.
```
[the remark this answers is not in the document]
Comment (pabloem) [reply website/www/site/data/capability_matrix.yaml:1713]: hm we have not stated this in Dataflow documentation (I'm working on that), but it's true for streaming.

**Review by pabloem — COMMENTED** [review]
I've finally gotten around to addressing your comments. Can you please take another look?

@kennknowles @je-ik @hughack

File: website/www/site/data/capability_matrix.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (hughack) [reply website/www/site/data/capability_matrix.yaml:1713]: Ok, i guess it makes sense to define the concept here first.

**Comment by hughack** [comment]
@pabloem - Yeah looks good, the only comment would be around @kennknowles point. I don't think "key-limited parallelism" is mentioned anywhere else in the Beam model. My understanding is it does happen, despite not explicitly being documented anywhere i can see. I'd be interested how Dataflow handles scaling the number of workers and if the guarantees still hold for any key that starts getting handled by a different worker.

**Review by je-ik — APPROVED** [review]
LGTM, but from higher level perspective, do we have any plans to support ordering regardless of the direct connection between the two PTransforms?

**Comment by pabloem** [comment]
@kennknowles PTAL : )

**Comment by pabloem** [comment]
@kennknowles PTAL - if no other comments by the end of the week, I'll merge form lazy consensus.

**Comment by pabloem** [comment]
moving forward as-is, but this is an area of active discussion, so please feel free to engage on this, and we cna work to keep this up to date.

**Comment by lukecwik** [comment]
https://github.com/apache/beam/pull/15378 broke spark PVR as well: https://ci-beam.apache.org/job/beam_PreCommit_Java_PVR_Spark_Batch_Cron/5140/ (build 5139 passes)

This Jenkins Job was renamed and is still [failing]( https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/32/console).

PVR tests sickbayed in https://github.com/apache/beam/pull/16411

**Comment by aaltay** [comment]
@pabloem @ibzib - Should this be reverted? (For reference https://issues.apache.org/jira/browse/BEAM-13522 is the tracker.)

**Comment by lukecwik** [comment]
I don't think it should be reverted but better care around adding new test categories without verifying them should be done in the future.

**Comment by ibzib** [comment]
I've proposed making Flink/Spark validates runner tests precommits for `runners/flink` and `runners/spark` file changes respectively, as well as `sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/` since often new VR tests are added there without checking if they pass on these runners. https://issues.apache.org/jira/browse/BEAM-13521

**Comment by aaltay** [comment]
I agree with @lukecwik, and @ibzib's proposal sounds like a good idea to enforce that.

@ibzib - How do we proceed with your proposal? Are you planning to do that? Are you waiting for input on the proposal?

**Comment by ibzib** [comment]
We've converted portable Flink VR to a precommit already, I'll follow up with other jobs later.

**Comment by aaltay** [comment]
> We've converted portable Flink VR to a precommit already, I'll follow up with other jobs later.

@ibzib, did the Spark VR also become a precommit?

**Comment by ibzib** [comment]
> > We've converted portable Flink VR to a precommit already, I'll follow up with other jobs later.
> 
> @ibzib, did the Spark VR also become a precommit?

See my comments on #16345

**Comment by aaltay** [comment]
> > > We've converted portable Flink VR to a precommit already, I'll follow up with other jobs later.
> > 
> > 
> > @ibzib, did the Spark VR also become a precommit?
> 
> See my comments on #16345

Ack. Thank you.
